### PR TITLE
Fix BatchNorm2d not realizing tensors

### PR DESCRIPTION
**Pull request by python273**
seem to fix `examples/train_resnet.py` getting slower over time

In test_mnist, there's `for p in model.parameters(): p.realize()` is some cases, but no in others. Should I just remove it, or there's some reason to not realize for some cases?

https://github.com/geohot/tinygrad/blob/e68fa18c9b3a4642ec2409102a8e4d0dd2a6f5bc/test/test_mnist.py#L51

---

lazy=0 is faster though (with torch=1)

LAZY=0
training: 5.44it/s
eval: 32.21it/s

LAZY=1
training: 4.29it/s
eval: 30.78it/s

### CODE REVIEW

**Comment by geohot** [comment]
That should be in the optimizer now, see nn/optim.py. Fine with removing it, though loss.realize() might still make sense.

I don't understand why this makes it slower though, realizing twice should be fine.

Slower over time = bug, but just slower in general might be because the MNIST models are so small. Try using the JIT.

**Comment by python273** [comment]
hmm, so `optim.step()` (calls `self.realize`) should be enough?

**Comment by python273** [comment]
hmm, seems like BatchNorm2d's tensors are not realized

- not realized
training 1.84it/s
eval 6.93s/it

- realized in `BatchNorm2d.__call__`
training 2.57it/s
eval 18.21it/s

https://github.com/geohot/tinygrad/blob/57aec117361fc35eb197659a8c7a19ee454eb865/tinygrad/nn/__init__.py#L10-L37

**Comment by python273** [comment]
saving stack when tensor is created for debugging:
https://gist.githubusercontent.com/python273/bd4d88387767b419a5e66b26e49da6a6/raw/16ab8f27d011e4840e64afafbdbe41d13179813d/tiny-tensors.txt

**Comment by geohot** [comment]
Ahh, yea running_mean and running_var are the absolute worst! Ideally they should just be turned off.

Saving the stack will long term be the way to build quick introspection tools into this.

**Comment by python273** [comment]
wdym turned off? 🤔

**Comment by python273** [comment]
can be merged I guess

**Comment by geohot** [comment]
Hmm, I don't really like this. I'm fine with realize on the __init__ (in fact, you might want to do .contiguous().realize() to make the underlying shapes correct for the weight and the bias)

But putting it on the others is realizing at the wrong time, and will cause the whole net to be realized. We need to find a way to delay it until the underlying tensor is ready.

**Comment by geohot** [comment]
Think I have a universal fix

**Comment by geohot** [comment]
https://github.com/geohot/tinygrad/commit/2f17d151b33dee7672e535fcbb930bfdc369bb96

BatchNorm seems broken though, test_conv_with_bn gets bad accuracy.

**Comment by geohot** [comment]
Yea, the LAZY=1 issue is fixed, but I just don't think batch norm training works.
